### DISCUSSION TITLE
Why receive is slow?

### DISCUSSION THREAD

**Discussion by rez0n** [AsamK/signal-cli#1412] [Q&A] [answered]
Hi, first of all thanks for this great tool!

I working on signal connector to external communications tool, so I need to implement two features
- **Send message** API endpoint, this is clear and works fine using `signal-cli -u xxx send -m 'text' uuid-of-receiver`.
- Automatically pool to **receive messages** `signal-cli -u xxx --output=json receive` - this feature takes something like 6 seconds to complete, which is means I could receive new messages something like every 10 seconds, which is sounds not cool.

Why receiving works so slow, is there any possibility to "listen" for incoming messages to make is close as possible to mobile app experience?

Thanks in advance.

**Comment by AsamK** [comment, marked as answer]
You probably want to use the `jsonRpc` or `daemon` command, where signal-cli keeps running and waits for messages from the server.

https://github.com/AsamK/signal-cli/blob/master/man/signal-cli.1.adoc#jsonrpc

https://github.com/AsamK/signal-cli/blob/master/man/signal-cli.1.adoc#daemon

**Reply by rez0n** [reply]
Much thanks, I used daemon and it works like a charm.

Below I leave simple example to make this discussion more useful for others.
How I start signal-cli daemon
```shell
signal-cli --config ~/.local/share/signal-cli -a '+xxxxxxxxxxx' daemon --http
```

How I listen for incoming messages in Python
```python
 api_url = 'http://127.0.0.1:8080/api/v1/events'
 with requests.get(api_url, headers=headers, stream=True) as response:
        for line in response.iter_lines():
            if line:
                decoded_line = line.decode('utf-8')
                if not decoded_line.startswith('data:'):
                    continue

                key, value = decoded_line.split(':', 1)
                event = json.loads(value.strip())
                event_data = event['envelope']
```

How I send message to Signal using signal-cli JsonRpc API endpoint
```python
    api_url = 'http://127.0.0.1:8080/api/v1/rpc'
    payload = {
        "jsonrpc": "2.0",
        "method": "send",
        "params":
            {
                "recipient": [recipient],
                "message": message,
            },
        "id": random.randint(10000, 99999)
    }
    
    headers = {
        "Content-Type": "application/json"
    }
    
    response = requests.post(api_url, data=json.dumps(payload), headers=headers)
```
